src/numeric/grids.jl

```diff
@@ -27,101 +27,165 @@ nodes(grid::EmptyGrid) = nothing
+#################
 
-    return UnstructuredGrid{N}(hcat(nodes...)')
+function mlinspace(min, max, n)
+    # this now returns an iterator
+    nodes = map(linspace, min, max, n)
+    return Base.product(nodes...)
 end
 
+immutable CartesianGrid{d} <: Grid{d}
+    min::Point{d}
+    max::Point{d}
+    n::SVector{d,Int}
+    nodes::ListOfPoints{d}
+end
 
-immutable CartesianGrid{N} <: Grid{N}
-    min::Vector{Float64}
-    max::Vector{Float64}
-    n::Vector{Int}
-    nodes::Matrix{Float64}
-    function (::Type{CartesianGrid{N}}){N}(min, max, n)
-        @assert N == length(min) == length(max) == length(n)
-        nodes = mlinspace(min, max, n)
-        return new{N}(min, max, n, nodes)
-    end
+function (::Type{<:CartesianGrid})(min::SVector{d,Float64}, max::SVector{d,Float64}, n::SVector{d,Int64}) where d
+    A = [mlinspace(min, max, n)...]
+    N = prod(n)
+    mm = reinterpret(Point{d},A,(N,))
+    return CartesianGrid{d}(min, max, n, mm)
 end
-nodes(grid::Grid) = grid.nodes
-n_nodes(grid::Grid) = size(grid.nodes, 1)
-node(grid::Grid, i::Int) = grid.nodes[i, :]
 
+(::Type{<:CartesianGrid})(min::Vector{Float64},max::Vector{Float64},n::Vector{Int64}) = CartesianGrid(SVector(min...), SVector(max...), SVector(n...))
+
+nodes(grid::CartesianGrid{d}) where d = grid.nodes
+n_nodes(grid::CartesianGrid{d}) where d = length(grid.nodes)
+node(grid::CartesianGrid{d}, i::Int) where d = grid.nodes[i]
 
-function Product(a::CartesianGrid, b::CartesianGrid)
-  N=length(a.min)+length(b.min)
-  return Dolo.CartesianGrid{N}(cat(1,a.min, b.min), cat(1, a.max, b.max), cat(1, a.n, b.n))
+nodes(::Type{<:ListOfPoints}, grid::CartesianGrid{d}) where d  = grid.nodes
+node(::Type{<:Point}, grid::CartesianGrid{d}, i::Int64) where d = node(grid,i)
+
+function Product(a::CartesianGrid{d1}, b::CartesianGrid{d2}) where d1 where d2
+  return Dolo.CartesianGrid{d1+d2}( [a.min; b.min], [a.max; b.max], [a.n; b.n])
 end
 
 
-immutable SmolyakGrid{N} <: Grid{N}
+################
+# Smolyak Grid #
+################
+
+immutable SmolyakGrid{d} <: Grid{d}
     smol_params::BM.SmolyakParams{Float64,Vector{Int}}
     nodes::Matrix{Float64}
     B_nodes::Matrix{Float64}
+end
 
-    function (::Type{SmolyakGrid{N}}){N}(min::Vector{Float64}, max::Vector{Float64}, mu::Vector{Int})
-        d = length(min)
-        @assert d == N
-        dim_err = DimensionMismatch("min was length $d, max and mu must match")
-        length(max) == d || throw(dim_err)
-        length(mu) == d || throw(dim_err)
-
-        sp = BM.SmolyakParams(d, mu, min, max)
-        nodes = BM.nodes(sp)
-        B_nodes = BM.evalbase(sp, nodes)
-        return new{N}(sp, nodes, B_nodes)
-    end
-    function (::Type{SmolyakGrid{N}}){N}(min::Array{Float64,1}, max::Array{Float64,1}, mu::Int)
-        return SmolyakGrid{N}(min, max, fill(mu, N))
-    end
+function SmolyakGrid{d}(min::Point{d}, max::Point{d}, mu::SVector{d,Int64}) where d
+    sp = BM.SmolyakParams(d, Vector(mu), Vector(min), Vector(max))
```

Comment:
Looks like some of our consistency checks in an inner constructor have been removed. Was that intentional?